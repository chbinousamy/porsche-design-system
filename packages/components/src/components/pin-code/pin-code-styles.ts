import type { FormState } from '../../utils/form/form-state';
import type { BreakpointCustomizable, Theme } from '../../types';
import { buildResponsiveStyles, getCss, isDisabledOrLoading, mergeDeep } from '../../utils';
import { getBaseChildStyles, getLabelStyles } from '../../styles/form-styles';
import { getFunctionalComponentRequiredStyles } from '../common/required/required-styles';
import { getFunctionalComponentStateMessageStyles } from '../common/state-message/state-message-styles';
import { getStylesWithoutSlottedSelector } from './pin-code-utils';
import { addImportantToEachRule, hostHiddenStyles } from '../../styles';
import {
  borderWidthBase,
  fontLineHeight,
  spacingStaticSmall,
  spacingStaticXSmall,
} from '@porsche-design-system/utilities-v2';

export const getComponentCss = (
  hideLabel: BreakpointCustomizable<boolean>,
  state: FormState,
  isDisabled: boolean,
  isLoading: boolean,
  isWithinForm: boolean,
  theme: Theme
): string => {
  const inputSize = `calc(${fontLineHeight} + 10px + ${borderWidthBase} * 2 + ${spacingStaticSmall} * 2)`; // we need 10px additionally so input height becomes 54px
  const labelStyles = getStylesWithoutSlottedSelector(
    getLabelStyles('input', isDisabledOrLoading(isDisabled, isLoading), hideLabel, state, theme)
  );
  const inputStyles = getStylesWithoutSlottedSelector(
    getBaseChildStyles('input', state, theme, {
      textAlign: 'center',
      width: inputSize,
      '@media (max-width: 480px)': {
        width: '39px', // max width that fits into grid
      },
      ...(isLoading && {
        opacity: 0.2,
      }),
    })
  );

  return getCss({
    '@global': {
      ':host': addImportantToEachRule({
        display: 'block',
        ...hostHiddenStyles,
      }),
      ...addImportantToEachRule({
        ...inputStyles,
        ...(isWithinForm && {
          '::slotted(input)': {
            position: 'absolute',
            height: inputSize,
            width: 0,
            opacity: 0,
          },
        }),
      }),
    },
    ...(isLoading && {
      spinner: {
        width: '100%',
        height: inputSize,
        pointerEvents: 'none',
        position: 'absolute',
        top: '50%',
        left: '50%',
        transform: 'translate(-50%, -50%)',
      },
    }),
    'pin-code-container': {
      display: 'flex',
      position: 'relative',
      gap: spacingStaticSmall,
      flexWrap: 'wrap',
      width: 'fit-content',
    },
    ...mergeDeep(labelStyles, {
      label: {
        ...buildResponsiveStyles(
          hideLabel,
          // workaround: since pin-code component is not wrapped into label tag it behaves differently
          (isLabelHidden: boolean) => (isLabelHidden ? { display: 'none' } : { display: 'inline-flex' })
        ),
        marginBottom: spacingStaticXSmall,
      },
    }),
    ...getFunctionalComponentRequiredStyles(),
    ...getFunctionalComponentStateMessageStyles(theme, state),
  });
};
