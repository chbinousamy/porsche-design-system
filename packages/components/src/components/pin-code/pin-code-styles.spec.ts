import { getComponentCss } from './pin-code-styles';

describe('getComponentCss()', () => {
  it.each<Parameters<typeof getComponentCss>>([
    [false, 'none', false, false, false, 'light'],
    [false, 'none', false, true, true, 'light'],
    [false, 'none', true, false, false, 'light'],
    [true, 'error', false, false, true, 'dark'],
    [{ base: true, xs: false, s: true, m: false, l: true, xl: false }, 'success', false, false, false, 'light'],
  ])(
    'should return correct css for hideLabel: %o, state: %s, isDisabled: %s, isLoading: %s, isWithinForm: %s and theme: %s',
    (...args) => {
      expect(getComponentCss(...args)).toMatchSnapshot();
    }
  );
});
