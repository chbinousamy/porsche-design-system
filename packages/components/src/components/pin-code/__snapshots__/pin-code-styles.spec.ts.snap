// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponentCss() should return correct css for hideLabel: { base: true, xs: false, s: true, m: false, l: true, xl: false }, state: success, isDisabled: false, isLoading: false, isWithinForm: false,theme: light 1`] = `
":host {
  display: block !important;
}
:host([hidden]) {
  display: none !important;
}
input {
  display: block !important;
  width: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  height: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  margin: 0 !important;
  outline: 0 !important;
  -webkit-appearance: none !important;
  appearance: none !important;
  box-sizing: border-box !important;
  border: 2px solid #197E10 !important;
  border-radius: 4px !important;
  background: transparent !important;
  font: normal normal 400 1rem/calc(6px + 2.125ex + 6px) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif !important;
  text-indent: 0 !important;
  color: #010205 !important;
  transition: color var(--p-transition-duration, .24s) ease,border-color var(--p-transition-duration, .24s) ease,background-color var(--p-transition-duration, .24s) ease !important;
  text-align: center !important;
}
input:focus {
  border-color: #010205 !important;
}
input:disabled {
  cursor: not-allowed !important;
  color: #949598 !important;
  border-color: #949598 !important;
  -webkit-text-fill-color: #949598 !important;
}
input[readonly] {
  border-color: #D8D8DB !important;
  background: #D8D8DB !important;
}
@media (max-width: 480px) {
  input {
    width: 39px !important;
  }
}

@media(hover:hover) {
  input:not(:disabled):not(:focus):not([readonly]):hover {
    border-color: #0E4809 !important;
  }
}

.pin-code-container {
  display: flex;
  position: relative;
  gap: 8px;
  flex-wrap: wrap;
  width: fit-content;
}
.label {
  display: none;
  flex-direction: column;
  gap: 4px;
  position: relative;
  margin-bottom: 4px;
}
.label__text {
  display: block;
  position: absolute;
  width: 0;
  height: 1px;
  text-indent: -999999px;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #010205;
  transition: color var(--p-transition-duration, .24s) ease;
}
.label__text+.label__text {
  margin-top: -4px;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
  color: #535457;
}
.required {
  user-select: none;
}
.message {
  display: flex;
  gap: 4px;
  margin-top: 4px;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #197E10;
  transition: color var(--p-transition-duration, .24s) ease;
}
@media(min-width:480px) {
  .label {
    display: inline-flex;
  }
  .label__text {
    position: static;
    width: fit-content;
    height: auto;
    text-indent: 0;
  }
}

@media(min-width:760px) {
  .label {
    display: none;
  }
  .label__text {
    position: absolute;
    width: 0;
    height: 1px;
    text-indent: -999999px;
  }
}

@media(min-width:1000px) {
  .label {
    display: inline-flex;
  }
  .label__text {
    position: static;
    width: fit-content;
    height: auto;
    text-indent: 0;
  }
}

@media(min-width:1300px) {
  .label {
    display: none;
  }
  .label__text {
    position: absolute;
    width: 0;
    height: 1px;
    text-indent: -999999px;
  }
}

@media(min-width:1760px) {
  .label {
    display: inline-flex;
  }
  .label__text {
    position: static;
    width: fit-content;
    height: auto;
    text-indent: 0;
  }
}

@media(hover:hover) {
  .label__text:hover~input:not(:disabled):not(:focus):not([readonly]), .label__text:hover input:not(:disabled):not(:focus):not([readonly]):hover {
    border-color: #0E4809 !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for hideLabel: false, state: none, isDisabled: false, isLoading: false, isWithinForm: false,theme: light 1`] = `
":host {
  display: block !important;
}
:host([hidden]) {
  display: none !important;
}
input {
  display: block !important;
  width: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  height: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  margin: 0 !important;
  outline: 0 !important;
  -webkit-appearance: none !important;
  appearance: none !important;
  box-sizing: border-box !important;
  border: 2px solid #6B6D70 !important;
  border-radius: 4px !important;
  background: transparent !important;
  font: normal normal 400 1rem/calc(6px + 2.125ex + 6px) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif !important;
  text-indent: 0 !important;
  color: #010205 !important;
  transition: color var(--p-transition-duration, .24s) ease,border-color var(--p-transition-duration, .24s) ease,background-color var(--p-transition-duration, .24s) ease !important;
  text-align: center !important;
}
input:focus {
  border-color: #010205 !important;
}
input:disabled {
  cursor: not-allowed !important;
  color: #949598 !important;
  border-color: #949598 !important;
  -webkit-text-fill-color: #949598 !important;
}
input[readonly] {
  border-color: #D8D8DB !important;
  background: #D8D8DB !important;
}
@media (max-width: 480px) {
  input {
    width: 39px !important;
  }
}

@media(hover:hover) {
  input:not(:disabled):not(:focus):not([readonly]):hover {
    border-color: #010205 !important;
  }
}

.pin-code-container {
  display: flex;
  position: relative;
  gap: 8px;
  flex-wrap: wrap;
  width: fit-content;
}
.label {
  display: inline-flex;
  flex-direction: column;
  gap: 4px;
  position: relative;
  margin-bottom: 4px;
}
.label__text {
  display: block;
  position: static;
  width: fit-content;
  height: auto;
  text-indent: 0;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #010205;
  transition: color var(--p-transition-duration, .24s) ease;
}
.label__text+.label__text {
  margin-top: -4px;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
  color: #535457;
}
.required {
  user-select: none;
}
.message {
  display: flex;
  gap: 4px;
  margin-top: 4px;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  transition: color var(--p-transition-duration, .24s) ease;
}
@media(hover:hover) {
  .label__text:hover~input:not(:disabled):not(:focus):not([readonly]) {
    border-color: #010205 !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for hideLabel: false, state: none, isDisabled: false, isLoading: true, isWithinForm: true,theme: light 1`] = `
":host {
  display: block !important;
}
:host([hidden]) {
  display: none !important;
}
input {
  display: block !important;
  width: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  height: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  margin: 0 !important;
  outline: 0 !important;
  -webkit-appearance: none !important;
  appearance: none !important;
  box-sizing: border-box !important;
  border: 2px solid #6B6D70 !important;
  border-radius: 4px !important;
  background: transparent !important;
  font: normal normal 400 1rem/calc(6px + 2.125ex + 6px) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif !important;
  text-indent: 0 !important;
  color: #010205 !important;
  transition: color var(--p-transition-duration, .24s) ease,border-color var(--p-transition-duration, .24s) ease,background-color var(--p-transition-duration, .24s) ease !important;
  text-align: center !important;
  opacity: 0.2 !important;
}
input:focus {
  border-color: #010205 !important;
}
input:disabled {
  cursor: not-allowed !important;
  color: #949598 !important;
  border-color: #949598 !important;
  -webkit-text-fill-color: #949598 !important;
}
input[readonly] {
  border-color: #D8D8DB !important;
  background: #D8D8DB !important;
}
::slotted(input) {
  position: absolute !important;
  height: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  width: 0 !important;
  opacity: 0 !important;
}
@media (max-width: 480px) {
  input {
    width: 39px !important;
  }
}

@media(hover:hover) {
  input:not(:disabled):not(:focus):not([readonly]):hover {
    border-color: #010205 !important;
  }
}

.spinner {
  width: 100%;
  height: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2);
  pointer-events: none;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}
.pin-code-container {
  display: flex;
  position: relative;
  gap: 8px;
  flex-wrap: wrap;
  width: fit-content;
}
.label {
  display: inline-flex;
  flex-direction: column;
  gap: 4px;
  position: relative;
  margin-bottom: 4px;
}
.label__text {
  display: block;
  position: static;
  width: fit-content;
  height: auto;
  text-indent: 0;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #949598;
  transition: color var(--p-transition-duration, .24s) ease;
}
.label__text+.label__text {
  margin-top: -4px;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
}
.required {
  user-select: none;
}
.message {
  display: flex;
  gap: 4px;
  margin-top: 4px;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  transition: color var(--p-transition-duration, .24s) ease;
}
@media(hover:hover) {
  .label__text:hover~input:not(:disabled):not(:focus):not([readonly]) {
    border-color: #010205 !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for hideLabel: false, state: none, isDisabled: true, isLoading: false, isWithinForm: false,theme: light 1`] = `
":host {
  display: block !important;
}
:host([hidden]) {
  display: none !important;
}
input {
  display: block !important;
  width: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  height: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  margin: 0 !important;
  outline: 0 !important;
  -webkit-appearance: none !important;
  appearance: none !important;
  box-sizing: border-box !important;
  border: 2px solid #6B6D70 !important;
  border-radius: 4px !important;
  background: transparent !important;
  font: normal normal 400 1rem/calc(6px + 2.125ex + 6px) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif !important;
  text-indent: 0 !important;
  color: #010205 !important;
  transition: color var(--p-transition-duration, .24s) ease,border-color var(--p-transition-duration, .24s) ease,background-color var(--p-transition-duration, .24s) ease !important;
  text-align: center !important;
}
input:focus {
  border-color: #010205 !important;
}
input:disabled {
  cursor: not-allowed !important;
  color: #949598 !important;
  border-color: #949598 !important;
  -webkit-text-fill-color: #949598 !important;
}
input[readonly] {
  border-color: #D8D8DB !important;
  background: #D8D8DB !important;
}
@media (max-width: 480px) {
  input {
    width: 39px !important;
  }
}

@media(hover:hover) {
  input:not(:disabled):not(:focus):not([readonly]):hover {
    border-color: #010205 !important;
  }
}

.pin-code-container {
  display: flex;
  position: relative;
  gap: 8px;
  flex-wrap: wrap;
  width: fit-content;
}
.label {
  display: inline-flex;
  flex-direction: column;
  gap: 4px;
  position: relative;
  margin-bottom: 4px;
}
.label__text {
  display: block;
  position: static;
  width: fit-content;
  height: auto;
  text-indent: 0;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #949598;
  transition: color var(--p-transition-duration, .24s) ease;
}
.label__text+.label__text {
  margin-top: -4px;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
}
.required {
  user-select: none;
}
.message {
  display: flex;
  gap: 4px;
  margin-top: 4px;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  transition: color var(--p-transition-duration, .24s) ease;
}
@media(hover:hover) {
  .label__text:hover~input:not(:disabled):not(:focus):not([readonly]) {
    border-color: #010205 !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for hideLabel: true, state: error, isDisabled: false, isLoading: false, isWithinForm: true,theme: dark 1`] = `
":host {
  display: block !important;
}
:host([hidden]) {
  display: none !important;
}
input {
  display: block !important;
  width: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  height: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  margin: 0 !important;
  outline: 0 !important;
  -webkit-appearance: none !important;
  appearance: none !important;
  box-sizing: border-box !important;
  border: 2px solid #FC4040 !important;
  border-radius: 4px !important;
  background: transparent !important;
  font: normal normal 400 1rem/calc(6px + 2.125ex + 6px) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif !important;
  text-indent: 0 !important;
  color: #FBFCFF !important;
  transition: color var(--p-transition-duration, .24s) ease,border-color var(--p-transition-duration, .24s) ease,background-color var(--p-transition-duration, .24s) ease !important;
  text-align: center !important;
}
input:focus {
  border-color: #FBFCFF !important;
}
input:disabled {
  cursor: not-allowed !important;
  color: #7E7F82 !important;
  border-color: #7E7F82 !important;
  -webkit-text-fill-color: #7E7F82 !important;
}
input[readonly] {
  border-color: #404044 !important;
  background: #404044 !important;
}
::slotted(input) {
  position: absolute !important;
  height: calc(calc(6px + 2.125ex) + 10px + 2px * 2 + 8px * 2) !important;
  width: 0 !important;
  opacity: 0 !important;
}
@media (max-width: 480px) {
  input {
    width: 39px !important;
  }
}

@media(hover:hover) {
  input:not(:disabled):not(:focus):not([readonly]):hover {
    border-color: #FB0404 !important;
  }
}

.pin-code-container {
  display: flex;
  position: relative;
  gap: 8px;
  flex-wrap: wrap;
  width: fit-content;
}
.label {
  display: none;
  flex-direction: column;
  gap: 4px;
  position: relative;
  margin-bottom: 4px;
}
.label__text {
  display: block;
  position: absolute;
  width: 0;
  height: 1px;
  text-indent: -999999px;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #FBFCFF;
  transition: color var(--p-transition-duration, .24s) ease;
}
.label__text+.label__text {
  margin-top: -4px;
  font-size: clamp(0.81rem, 0.23vw + 0.77rem, 0.88rem);
  color: #AFB0B3;
}
.required {
  user-select: none;
}
.message {
  display: flex;
  gap: 4px;
  margin-top: 4px;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: break-word;
  hyphens: auto;
  color: #FC4040;
  transition: color var(--p-transition-duration, .24s) ease;
}
@media(hover:hover) {
  .label__text:hover~input:not(:disabled):not(:focus):not([readonly]), .label__text:hover input:not(:disabled):not(:focus):not([readonly]):hover {
    border-color: #FB0404 !important;
  }
}
"
`;
